Parallel testing on Python >= 3.3 breaks due to pseudo-nondeterministic test collection

I will post the full output below, but the relevant portion is this:

```
INTERNALERROR> AssertionError: Different tests were collected between gw5 and gw3. The difference is:
INTERNALERROR> --- gw5
INTERNALERROR> 
INTERNALERROR> +++ gw3
INTERNALERROR> 
INTERNALERROR> @@ -2693,7 +2693,7 @@
INTERNALERROR> 
INTERNALERROR>  astropy/table/np_utils.py::astropy.table.np_utils.vstack
INTERNALERROR>  astropy/table/operations.py::astropy.table.operations.hstack
INTERNALERROR>  astropy/table/operations.py::astropy.table.operations.vstack
INTERNALERROR> -astropy/table/table.py::astropy.table.table.Row
INTERNALERROR> +astropy/table/table.py::astropy.table.table.Table.Row
INTERNALERROR>  astropy/table/table.py::astropy.table.table.Table.add_column
INTERNALERROR>  astropy/table/table.py::astropy.table.table.Table.add_columns
INTERNALERROR>  astropy/table/table.py::astropy.table.table.Table.add_row
```

I think that what's happening here is that `astropy/table/table.py::astropy.table.table.Row` and `astropy/table/table.py::astropy.table.table.Table.Row` both refer to the same object.  But for some reason one test runner is finding the latter before the former.  I'm going ahead and marking this "upstream fix required" because the test collector should be able to recognize those as the same object, though I'm not sure about the details.

This might also not be just on Python 3.4, but that was the only Python version where _I_ could reproduce the issue.  It's probably something pseudo-nondeterminisitic.

```
====================================================== test session starts =======================================================
platform linux -- Python 3.4.0 -- pytest-2.5.1

Running tests with Astropy version 1.0.dev9659.
Running tests in /tmp/astropy-test-moymwonq/lib.linux-x86_64-3.4/astropy /internal/1/root/src/astropy/astropy/docs.

Platform: Linux-2.6.32-431.20.3.el6.x86_64-x86_64-with-redhat-6.5-Santiago

Executable: /internal/1/root/bin/overrides/python3.4

Full Python Version: 
3.4.0 (default, Apr 29 2014, 16:13:55) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-4)]

encodings: sys: utf-8, locale: UTF-8, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15

Numpy: 1.9.0.dev-7f46687
Scipy: 0.13.3
Matplotlib: not available
h5py: not available
Running doctests in .rst files is not supported on Python 3.x

plugins: xdist, cov
gw0 [5386] / gw1 [5386] / gw2 [5386] / gw3 [5386] / gw4 [5386] / gw5 [5386]
scheduling tests via LoadScheduling
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "_pytest.main", line 81, in wrap_session
INTERNALERROR>     doit(config, session)
INTERNALERROR>   File "_pytest.main", line 117, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "_pytest.core", line 376, in __call__
INTERNALERROR>     return self._docall(methods, kwargs)
INTERNALERROR>   File "_pytest.core", line 387, in _docall
INTERNALERROR>     res = mc.execute()
INTERNALERROR>   File "_pytest.core", line 288, in execute
INTERNALERROR>     res = method(**kwargs)
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/xdist/dsession.py", line 244, in pytest_runtestloop
INTERNALERROR>     self.loop_once()
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/xdist/dsession.py", line 262, in loop_once
INTERNALERROR>     call(**kwargs)
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/xdist/dsession.py", line 315, in slave_collectionfinish
INTERNALERROR>     self.sched.init_distribute()
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/xdist/dsession.py", line 138, in init_distribute
INTERNALERROR>     node.gateway.id,
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/xdist/dsession.py", line 188, in report_collection_diff
INTERNALERROR>     raise AssertionError(msg)
INTERNALERROR> AssertionError: Different tests were collected between gw5 and gw3. The difference is:
INTERNALERROR> --- gw5
INTERNALERROR> 
INTERNALERROR> +++ gw3
INTERNALERROR> 
INTERNALERROR> @@ -2693,7 +2693,7 @@
INTERNALERROR> 
INTERNALERROR>  astropy/table/np_utils.py::astropy.table.np_utils.vstack
INTERNALERROR>  astropy/table/operations.py::astropy.table.operations.hstack
INTERNALERROR>  astropy/table/operations.py::astropy.table.operations.vstack
INTERNALERROR> -astropy/table/table.py::astropy.table.table.Row
INTERNALERROR> +astropy/table/table.py::astropy.table.table.Table.Row
INTERNALERROR>  astropy/table/table.py::astropy.table.table.Table.add_column
INTERNALERROR>  astropy/table/table.py::astropy.table.table.Table.add_columns
INTERNALERROR>  astropy/table/table.py::astropy.table.table.Table.add_row
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/main.py", line 81, in wrap_session
INTERNALERROR>     doit(config, session)
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/main.py", line 118, in _main
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/main.py", line 81, in wrap_session
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/core.py", line 377, in __call__
INTERNALERROR>     doit(config, session)
INTERNALERROR>     return self._docall(methods, kwargs)
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/main.py", line 118, in _main
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/core.py", line 388, in _docall
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>     res = mc.execute()
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/core.py", line 377, in __call__
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/core.py", line 289, in execute
INTERNALERROR>     return self._docall(methods, kwargs)
INTERNALERROR>     res = method(**kwargs)
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/core.py", line 388, in _docall
INTERNALERROR>   File "<remote exec>", line 47, in pytest_runtestloop
INTERNALERROR>     res = mc.execute()
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/execnet/gateway_base.py", line 701, in receive
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/core.py", line 289, in execute
INTERNALERROR>     raise self._getremoteerror() or EOFError()
INTERNALERROR>     res = method(**kwargs)
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/main.py", line 81, in wrap_session
INTERNALERROR> EOFError
INTERNALERROR>   File "<remote exec>", line 47, in pytest_runtestloop
INTERNALERROR>     doit(config, session)
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/execnet/gateway_base.py", line 701, in receive
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/main.py", line 118, in _main
INTERNALERROR>     raise self._getremoteerror() or EOFError()
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR> EOFError
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/core.py", line 377, in __call__
INTERNALERROR>     return self._docall(methods, kwargs)
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/core.py", line 388, in _docall
INTERNALERROR>     res = mc.execute()
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/core.py", line 289, in execute
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>     res = method(**kwargs)
INTERNALERROR>   File "<remote exec>", line 47, in pytest_runtestloop
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/main.py", line 81, in wrap_session
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/execnet/gateway_base.py", line 701, in receive
INTERNALERROR>     doit(config, session)
INTERNALERROR>     raise self._getremoteerror() or EOFError()
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/main.py", line 118, in _main
INTERNALERROR> EOFError
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/core.py", line 377, in __call__
INTERNALERROR>     return self._docall(methods, kwargs)
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/core.py", line 388, in _docall
INTERNALERROR>     res = mc.execute()
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/core.py", line 289, in execute
INTERNALERROR>     res = method(**kwargs)
INTERNALERROR>   File "<remote exec>", line 47, in pytest_runtestloop
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/execnet/gateway_base.py", line 701, in receive
INTERNALERROR>     raise self._getremoteerror() or EOFError()
INTERNALERROR> EOFError
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/main.py", line 81, in wrap_session
INTERNALERROR>     doit(config, session)
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/main.py", line 118, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/core.py", line 377, in __call__
INTERNALERROR>     return self._docall(methods, kwargs)
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/core.py", line 388, in _docall
INTERNALERROR>     res = mc.execute()
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/core.py", line 289, in execute
INTERNALERROR>     res = method(**kwargs)
INTERNALERROR>   File "<remote exec>", line 47, in pytest_runtestloop
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/execnet/gateway_base.py", line 701, in receive
INTERNALERROR>     raise self._getremoteerror() or EOFError()
INTERNALERROR> EOFError
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/main.py", line 81, in wrap_session
INTERNALERROR>     doit(config, session)
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/main.py", line 118, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/core.py", line 377, in __call__
INTERNALERROR>     return self._docall(methods, kwargs)
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/core.py", line 388, in _docall
INTERNALERROR>     res = mc.execute()
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/_pytest/core.py", line 289, in execute
INTERNALERROR>     res = method(**kwargs)
INTERNALERROR>   File "<remote exec>", line 47, in pytest_runtestloop
INTERNALERROR>   File "/internal/1/root/usr/local/lib/python3.4/site-packages/execnet/gateway_base.py", line 701, in receive
INTERNALERROR>     raise self._getremoteerror() or EOFError()
INTERNALERROR> EOFError
sys:1: ResourceWarning: unclosed file <_io.TextIOWrapper name='/tmp/astropy-test-moymwonq/lib.linux-x86_64-3.4/astropy/extern/bundled/six.py' mode='r' encoding='utf-8'>
sys:1: ResourceWarning: unclosed file <_io.TextIOWrapper name='/tmp/astropy-test-moymwonq/lib.linux-x86_64-3.4/astropy/extern/bundled/six.py' mode='r' encoding='utf-8'>
sys:1: ResourceWarning: unclosed file <_io.TextIOWrapper name='/tmp/astropy-test-moymwonq/lib.linux-x86_64-3.4/astropy/extern/bundled/six.py' mode='r' encoding='utf-8'>
```

--- a/astropy/table/row.py
+++ b/astropy/table/row.py
@@ -1,0 +1,150 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+
+import collections
+from distutils import version
+
+import numpy as np
+from numpy import ma
+
+from ..extern import six
+
+class Row(object):
+    """A class to represent one row of a Table object.
+
+    A Row object is returned when a Table object is indexed with an integer
+    or when iterating over a table::
+
+      >>> from astropy.table import Table
+      >>> table = Table([(1, 2), (3, 4)], names=('a', 'b'),
+      ...               dtype=('int32', 'int32'))
+      >>> row = table[1]
+      >>> row
+      <Row 1 of table
+       values=(2, 4)
+       dtype=[('a', '<i4'), ('b', '<i4')]>
+      >>> row['a']
+      2
+      >>> row[1]
+      4
+    """
+
+    def __init__(self, table, index):
+        self._table = table
+        self._index = index
+        try:
+            self._data = table._data[index]
+
+            # MaskedArray __getitem__ has a strange behavior where if a
+            # row mask is all False then it returns a np.void which
+            # has no mask attribute. This makes it impossible to then set
+            # the mask. Here we recast back to mvoid. This was fixed in
+            # Numpy following issue numpy/numpy#483, and the fix should be
+            # included in Numpy 1.8.0.
+            if self._table.masked and isinstance(self._data, np.void):
+                self._data = ma.core.mvoid(self._data,
+                                           mask=self._table._mask[index])
+        except ValueError as err:
+            # Another bug (or maybe same?) that is fixed in 1.8 prevents
+            # accessing a row in masked array if it has object-type members.
+            # >>> x = np.ma.empty(1, dtype=[('a', 'O')])
+            # >>> x['a'] = 1
+            # >>> x['a'].mask = True
+            # >>> x[0]
+            # ValueError: Setting void-array with object members using buffer. [numpy.ma.core]
+            #
+            # All we do here is re-raise with a more informative message
+            if (six.text_type(err).startswith('Setting void-array with object members')
+                    and version.LooseVersion(np.__version__) < version.LooseVersion('1.8')):
+                raise ValueError('Cannot access table row with Object type columns, due to '
+                                 'a bug in numpy {0}.  Please upgrade to numpy 1.8 or newer.'
+                                 .format(np.__version__))
+            else:
+                raise
+
+    def __getitem__(self, item):
+        return self.data[item]
+
+    def __setitem__(self, item, val):
+        if self._table.masked:
+            # Workaround for astropy/astropy#2734 and numpy/numpy#4866:
+            # Assignment to a masked array containing a recarray object doesn't
+            # work properly when being assigned in [row][colname] order.
+            # Instead go back to the parent table and do [colname][row] order.
+            #
+            # Note also that in the masked case the Row object is not a direct
+            # view of the data so we need to set in the table and in self.data.
+            col = self._table.columns[item]  # works for index or col name
+            col[self._index] = val
+
+        self.data[item] = val
+
+    def __eq__(self, other):
+        if self._table.masked:
+            # Sent bug report to numpy-discussion group on 2012-Oct-21, subject:
+            # "Comparing rows in a structured masked array raises exception"
+            # No response, so this is still unresolved.
+            raise ValueError('Unable to compare rows for masked table due to numpy.ma bug')
+        return self.data == other
+
+    def __ne__(self, other):
+        if self._table.masked:
+            raise ValueError('Unable to compare rows for masked table due to numpy.ma bug')
+        return self.data != other
+
+    @property
+    def _mask(self):
+        return self._data.mask
+
+    def __array__(self, dtype=None):
+        """Support converting Row to np.array via np.array(table).
+
+        Coercion to a different dtype via np.array(table, dtype) is not
+        supported and will raise a ValueError.
+        """
+        if dtype is not None:
+            raise ValueError('Datatype coercion is not allowed')
+
+        return np.array(self._data)
+
+    def __len__(self):
+        return len(self._data.dtype)
+
+    @property
+    def table(self):
+        return self._table
+
+    @property
+    def index(self):
+        return self._index
+
+    @property
+    def data(self):
+        return self._data
+
+    @property
+    def meta(self):
+        return self.table.meta
+
+    @property
+    def columns(self):
+        return self.table.columns
+
+    @property
+    def colnames(self):
+        return self.table.colnames
+
+    @property
+    def dtype(self):
+        return self.data.dtype
+
+    def __repr__(self):
+        return "<{3} {0} of table\n values={1!r}\n dtype={2}>".format(
+            self.index, self.data, self.dtype, self.__class__.__name__)
+
+
+collections.Sequence.register(Row)
+
+

--- a/astropy/table/table.py
+++ b/astropy/table/table.py
@@ -5,7 +5,6 @@ from ..extern import six
 from ..extern.six.moves import zip as izip
 from ..extern.six.moves import range as xrange
 
-import collections
 import warnings
 import re
 
@@ -25,6 +24,7 @@ from ..utils.metadata import MetaData
 from . import groups
 from .pprint import TableFormatter
 from .column import BaseColumn, Column, MaskedColumn, _auto_names
+from .row import Row
 from .np_utils import fix_column_name
 
 
@@ -107,142 +107,6 @@ class TableColumns(OrderedDict):
 
     def values(self):
         return list(OrderedDict.values(self))
-
-
-class Row(object):
-    """A class to represent one row of a Table object.
-
-    A Row object is returned when a Table object is indexed with an integer
-    or when iterating over a table::
-
-      >>> table = Table([(1, 2), (3, 4)], names=('a', 'b'),
-      ...               dtype=('int32', 'int32'))
-      >>> row = table[1]
-      >>> row
-      <Row 1 of table
-       values=(2, 4)
-       dtype=[('a', '<i4'), ('b', '<i4')]>
-      >>> row['a']
-      2
-      >>> row[1]
-      4
-    """
-
-    def __init__(self, table, index):
-        self._table = table
-        self._index = index
-        try:
-            self._data = table._data[index]
-
-            # MaskedArray __getitem__ has a strange behavior where if a
-            # row mask is all False then it returns a np.void which
-            # has no mask attribute. This makes it impossible to then set
-            # the mask. Here we recast back to mvoid. This was fixed in
-            # Numpy following issue numpy/numpy#483, and the fix should be
-            # included in Numpy 1.8.0.
-            if self._table.masked and isinstance(self._data, np.void):
-                self._data = ma.core.mvoid(self._data,
-                                           mask=self._table._mask[index])
-        except ValueError as err:
-            # Another bug (or maybe same?) that is fixed in 1.8 prevents
-            # accessing a row in masked array if it has object-type members.
-            # >>> x = np.ma.empty(1, dtype=[('a', 'O')])
-            # >>> x['a'] = 1
-            # >>> x['a'].mask = True
-            # >>> x[0]
-            # ValueError: Setting void-array with object members using buffer. [numpy.ma.core]
-            #
-            # All we do here is re-raise with a more informative message
-            if (six.text_type(err).startswith('Setting void-array with object members')
-                    and version.LooseVersion(np.__version__) < version.LooseVersion('1.8')):
-                raise ValueError('Cannot access table row with Object type columns, due to '
-                                 'a bug in numpy {0}.  Please upgrade to numpy 1.8 or newer.'
-                                 .format(np.__version__))
-            else:
-                raise
-
-    def __getitem__(self, item):
-        return self.data[item]
-
-    def __setitem__(self, item, val):
-        if self._table.masked:
-            # Workaround for astropy/astropy#2734 and numpy/numpy#4866:
-            # Assignment to a masked array containing a recarray object doesn't
-            # work properly when being assigned in [row][colname] order.
-            # Instead go back to the parent table and do [colname][row] order.
-            #
-            # Note also that in the masked case the Row object is not a direct
-            # view of the data so we need to set in the table and in self.data.
-            col = self._table.columns[item]  # works for index or col name
-            col[self._index] = val
-
-        self.data[item] = val
-
-    def __eq__(self, other):
-        if self._table.masked:
-            # Sent bug report to numpy-discussion group on 2012-Oct-21, subject:
-            # "Comparing rows in a structured masked array raises exception"
-            # No response, so this is still unresolved.
-            raise ValueError('Unable to compare rows for masked table due to numpy.ma bug')
-        return self.data == other
-
-    def __ne__(self, other):
-        if self._table.masked:
-            raise ValueError('Unable to compare rows for masked table due to numpy.ma bug')
-        return self.data != other
-
-    @property
-    def _mask(self):
-        return self._data.mask
-
-    def __array__(self, dtype=None):
-        """Support converting Row to np.array via np.array(table).
-
-        Coercion to a different dtype via np.array(table, dtype) is not
-        supported and will raise a ValueError.
-        """
-        if dtype is not None:
-            raise ValueError('Datatype coercion is not allowed')
-
-        return np.array(self._data)
-
-    def __len__(self):
-        return len(self._data.dtype)
-
-    @property
-    def table(self):
-        return self._table
-
-    @property
-    def index(self):
-        return self._index
-
-    @property
-    def data(self):
-        return self._data
-
-    @property
-    def meta(self):
-        return self.table.meta
-
-    @property
-    def columns(self):
-        return self.table.columns
-
-    @property
-    def colnames(self):
-        return self.table.colnames
-
-    @property
-    def dtype(self):
-        return self.data.dtype
-
-    def __repr__(self):
-        return "<{3} {0} of table\n values={1!r}\n dtype={2}>".format(
-            self.index, self.data, self.dtype, self.__class__.__name__)
-
-
-collections.Sequence.register(Row)
 
 
 class Table(object):

